test/dynamo/test_metrics_context.py

```diff
@@ -0,0 +1,59 @@
+# Owner(s): ["module: dynamo"]
+
+from torch._dynamo.metrics_context import MetricsContext
+from torch._dynamo.test_case import run_tests, TestCase
+
+
+class TestMetricsContext(TestCase):
+    def setUp(self):
+        super().setUp()
+        self.metrics = {}
+
+    def _on_exit(self, metrics, exc_type, exc_value):
+        # Save away the metrics to be validated in the test.
+        self.metrics = metrics.copy()
+
+    def test_context_exists(self):
+        """
+        Setting a value without entering the context should raise.
+        """
+        context = MetricsContext(self._on_exit)
+        with self.assertRaisesRegex(RuntimeError, "outside of a MetricsContext"):
+            context.increment("m", 1)
+
+        with self.assertRaisesRegex(RuntimeError, "outside of a MetricsContext"):
+            context.set("m", 1)
+
+        with self.assertRaisesRegex(RuntimeError, "outside of a MetricsContext"):
+            context.set_once("m", 1)
+
+        with self.assertRaisesRegex(RuntimeError, "outside of a MetricsContext"):
+            context.update({"m", 1})
+
+    def test_nested_context(self):
+        """
+        Only the outermost context should get an on_exit call, and it should
+        include everything.
+        """
+        context = MetricsContext(self._on_exit)
+        with context:
+            with context:
+                context.set("m1", 1)
+            self.assertEqual(self.metrics, {})
+            context.set("m2", 2)
+        self.assertEqual(self.metrics, {"m1": 1, "m2": 2})
+
+    def test_set(self):
+        """
+        Validate various ways to set metrics.
+        """
+        with MetricsContext(self._on_exit) as context:
+            context.set("m1", 1)
+            context.set_once("m2", 2)
+            context.update({"m3": 3, "m4": 4})
+
+        self.assertEqual(self.metrics, {"m1": 1, "m2": 2, "m3": 3, "m4": 4})
```

Comment:
Yeah, more tests that make it clear what aggregation semantics are would be helpful.

Reply:
I've changed the API to have an explicit overwrite flag (Default to False). Hopefully that makes it more clear. Also new tests below.

Reply:
fwiw you don't actually use overwrite ;)

Reply:
(It's not clear to me when you would ever want to use overwrite)

Reply:
removing